flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java

```diff
@@ -0,0 +1,179 @@
+ *      |            |       /resource_manager/connection_info
+ *      |            |       /dispatcher/connection_info
+ *      |            |       /rest_server/connection_info
+ *      |            |       /job-id-1/connection_info
+ *      |            |       /job-id-2/connection_info
+ *      |            |
+ *      |            |
+ *      |            +jobgraphs/job-id-1
+ *      |            |         /job-id-2
+ *      |            +jobs/job-id-1/checkpoints/latest
+ *      |                 |                    /latest-1
+ *      |                 |                    /latest-2
+ *      |                 |       /checkpoint_id_counter
+ * </pre>
+ */
+public class ZooKeeperMultipleComponentLeaderElectionHaServices
+        extends AbstractZooKeeperHaServices {
+
+    private final Object lock = new Object();
+
+    private final CuratorFramework leaderNamespacedCuratorFramework;
+
+    private final FatalErrorHandler fatalErrorHandler;
+
+    @Nullable
+    @GuardedBy("lock")
+    private MultipleComponentLeaderElectionService multipleComponentLeaderElectionService = null;
+
+    public ZooKeeperMultipleComponentLeaderElectionHaServices(
+            CuratorFrameworkWithUnhandledErrorListener curatorFrameworkWrapper,
+            Configuration config,
+            Executor ioExecutor,
+            BlobStoreService blobStoreService,
+            FatalErrorHandler fatalErrorHandler)
+            throws Exception {
+        super(curatorFrameworkWrapper, ioExecutor, config, blobStoreService);
+        this.leaderNamespacedCuratorFramework =
+                ZooKeeperUtils.useNamespaceAndEnsurePath(
+                        getCuratorFramework(), ZooKeeperUtils.getLeaderPath());
+        this.fatalErrorHandler = fatalErrorHandler;
+    }
+
+    @Override
+    protected LeaderElectionService createLeaderElectionService(String leaderName) {
+        final MultipleComponentLeaderElectionService multipleComponentLeaderElectionService;
+
+        synchronized (lock) {
+            multipleComponentLeaderElectionService = getOrInitializeSingleLeaderElectionService();
+        }
+
+        return new DefaultLeaderElectionService(
+                multipleComponentLeaderElectionService.createDriverFactory(leaderName));
+    }
+
+    @GuardedBy("lock")
+    private MultipleComponentLeaderElectionService getOrInitializeSingleLeaderElectionService() {
+        if (multipleComponentLeaderElectionService == null) {
+            try {
+                multipleComponentLeaderElectionService =
+                        new DefaultMultipleComponentLeaderElectionService(
+                                fatalErrorHandler,
+                                "Single leader election service.",
+                                new ZooKeeperMultipleComponentLeaderElectionDriverFactory(
+                                        leaderNamespacedCuratorFramework));
+            } catch (Exception e) {
+                throw new FlinkRuntimeException(
```

Comment:
Yes, the `fatalErrorHandler` is used for asynchronous exception that might occur. When doing synchronous call, I think it is better to directly throw the exception so that the caller can handle it appropriately.